parse.py

```diff
@@ -138,55 +137,94 @@ def query_pageids(pageids):
             content = page['revisions'][0]['slots']['main']['content']
             yield (revid, title, content)
 
-
-def parse(pageids):
-    page_processed = 0
-    pageids_list = list(pageids) # for test, sholud upgrade to multi-threads version
-    model, vocab_w2v, section_dict = load_citation_needed()
+def with_max_exceptions(fn):
+    @functools.wraps(fn)
+    def wrapper(*args, **kwds):
+        try:
+            return fn(*args, **kwds)
+        except:
+            traceback.print_exc()
+            self.exception_count += 1
+            if self.exception_count > MAX_EXCEPTIONS_PER_SUBPROCESS:
+                print('Too many exceptions, quitting!')
+                raise
+    return wrapper
+
+@with_max_exceptions
+def work(model, pageids):
+    rows = []
+    results = query_pageids(pageids)
+    for revid, title, wikitext in results:
+        start_len = len(rows)
+        wikicode = mwparserfromhell.parse(wikitext)
+        sentences, paragraphs = extract(wikicode)
+
+        for text, pidx, section in sentences:
+            row = [text, paragraphs[pidx], section, revid]
+            rows.append(row)
+
+        X, sections = run_citation_needed(sentences)
+        pred = model.predict([X, sections])
+        for i, score in enumerate(pred):
+            rows[start_len+i].append(score[1])
 
     def insert(cursor, r):
         cursor.execute('''
             INSERT INTO statements (statement, context, section, rev_id, score)
             VALUES(%s, %s, %s, %s, %s)
             ''', r)
     db = cddb.init_scratch_db()
+    for r in rows:
+        try:
+            db.execute_with_retry(insert, r)
+        except:
+            # This may be caused by the size
+            # of a sentence or paragraph that
+            # exceed the maximum limit
+            continue
 
-    batch_size = 32
-    for i in range(0, len(pageids_list), batch_size):
-        rows = []
-        results = query_pageids(pageids_list[i: i + batch_size])
-        for revid, title, wikitext in results:
-            start_len = len(rows)
-            wikicode = mwparserfromhell.parse(wikitext)
-            sentences, paragraphs = extract(wikicode)
-
-            for text, pidx, section in sentences:
-                row = [text, paragraphs[pidx], section, revid]
-                rows.append(row)
-
-            pred = run_citation_needed(sentences, model, vocab_w2v, section_dict)
-            for i, score in enumerate(pred):
-                rows[start_len+i].append(score[1])
+def parse(pageids, timeout):
+    from keras_model import KerasManager
+    manager = KerasManager()
+    manager.start()
+    model = manager.KerasModel()
+    model.initialize()
 
-            page_processed += 1
+    pool = multiprocessing.Pool(
+        processes = multiprocessing.cpu_count(), 
```

Comment:
`cpu_count()` is actually the default, so we don't need to specify it.

By the way, I'd expect that these processes will spend most of their time waiting on the Wikipedia API rather than doing any processing, which means we could also be spawning more processes than we have CPUs and the OS will just re-schedule them as needed. See here: https://github.com/eggpi/citationhunt/blob/master/scripts/parse_live.py#L186